iio/iio_ad7124/iio_ad7124.c

```diff
@@ -0,0 +1,1061 @@
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+static ssize_t ad1724_iio_read_scale_chan(void *device, char *buf, size_t len,
+		const struct iio_ch_info *channel)
+{
+	struct ad7124_dev	*desc = (struct ad7124_dev *)device;
+	uint32_t		vref_mv = 2500, adc_bit_no = 24;
+	uint32_t		pga_bits, bipolar;
+	float			lsb_val;
+
+	ad7124_read_register(desc, &desc->regs[AD7124_CFG0_REG + channel->ch_num]);
+	pga_bits = desc->regs[AD7124_CFG0_REG + channel->ch_num].value & 0x7;
+	bipolar = (desc->regs[AD7124_CFG0_REG + channel->ch_num].value >> 11) & 0x1;
+	if (bipolar != 0)
+		lsb_val = vref_mv / pow(2, (adc_bit_no + pga_bits - 1));
+	else
+		lsb_val = vref_mv / pow(2, (adc_bit_no + pga_bits));
+
+	return snprintf(buf, len, "%f", lsb_val);
+}
+
+/**
+ * @brief Find closest possible channel gain from a free input one.
+ * @param [in] new_gain - Input gain.
+ * @return New PGA setting based on the gain.
+ */
+static uint32_t ad7124_iio_find_closest_gain(uint32_t new_gain)
+{
+	uint32_t new_pga_bits;
+	uint32_t old_diff = 0xFFFFFFFF, new_diff;
+	uint32_t i;
+
+	for (i = 0; i < 8; i++) {
+		new_diff = abs(pow(2, i) - new_gain);
+		if (new_diff < old_diff) {
+			old_diff = new_diff;
+			new_pga_bits = i;
+		}
+	}
+
+	return new_pga_bits;
+}
+
+/**
+ * @brief Change the channel LSB voltage value.
+ * @param device - Device driver descriptor.
+ * @param buf - Input buffer.
+ * @param len - Length of the input buffer (not used in this case).
+ * @param channel - IIO channel information.
+ * @return Number of bytes printed in the output buffer, or negative error code.
+ */
+static ssize_t ad1724_iio_change_scale_chan(void *device, char *buf, size_t len,
+		const struct iio_ch_info *channel)
+{
+	struct ad7124_dev	*desc = (struct ad7124_dev *)device;
+	float			new_scale;
+	uint32_t		vref_mv = 2500, adc_bit_no = 24;
+	uint32_t		bipolar;
+	float			lsb_val;
+	uint32_t		new_gain;
+
+	sscanf(buf, "%f", &new_scale);
+
+	ad7124_read_register(desc, &desc->regs[AD7124_CFG0_REG + channel->ch_num]);
+	bipolar = (desc->regs[AD7124_CFG0_REG + channel->ch_num].value >> 11) & 0x1;
+	if (bipolar != 0)
+		lsb_val = vref_mv / pow(2, (adc_bit_no - 1));
+	else
+		lsb_val = vref_mv / pow(2, adc_bit_no);
+
+	new_gain = lsb_val / new_scale;
+
+	new_gain = ad7124_iio_find_closest_gain(new_gain);
+	desc->regs[AD7124_CFG0_REG + channel->ch_num].value &= ~AD7124_CFG_REG_PGA(~0);
+	desc->regs[AD7124_CFG0_REG + channel->ch_num].value |= AD7124_CFG_REG_PGA(
+				new_gain);
+	ad7124_write_register(desc, desc->regs[AD7124_CFG0_REG + channel->ch_num]);
+
+	return len;
+}
+
+/**
+ * @brief Transfer data from device into RAM.
+ * @param iio_inst - Physical instance of a iio_axi_adc device.
+ * @param bytes_count - Number of bytes to transfer.
+ * @param ch_mask - Opened channels mask.
+ * @return bytes_count or negative value in case of error.
+ */
+static ssize_t iio_ad7124_transfer_dev_to_mem(void *iio_inst,
```

Comment:
Did you tried: https://github.com/analogdevicesinc/no-OS/blob/5a93b5a6e9fb132fd92a1bef0050eb27d5e32aa8/iio/iio_demo/iio_demo_dev.h#L151

Reply:
and read_dev functions